File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/myshop/usecases/BrowseCoupons.java
package com.mycompany.myshop.testkit.dsl.core.usecase.myshop.usecases;

import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.usecase.myshop.usecases.base.BaseMyShopUseCase;
import com.mycompany.myshop.testkit.driver.port.myshop.MyShopDriver;
import com.mycompany.myshop.testkit.driver.port.myshop.dtos.BrowseCouponsResponse;

public class BrowseCoupons extends BaseMyShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification> {
    public BrowseCoupons(MyShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    @Override
    public UseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification> execute() {
        var result = driver.browseCoupons();
        return new UseCaseResult<>(result, context, BrowseCouponsVerification::new);
    }
}

File: system-test/dotnet/Dsl.Core/UseCase/MyShop/UseCases/BrowseCoupons.cs
using Driver.Port.MyShop;
using Driver.Port.MyShop.Dtos;
using Driver.Port.MyShop.Dtos.Error;
using Dsl.Core.Shared;
using Dsl.Core.MyShop.UseCases.Base;

namespace Dsl.Core.MyShop.UseCases;

public class BrowseCoupons : BaseMyShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification>
{
    public BrowseCoupons(IMyShopDriver driver, UseCaseContext context)
        : base(driver, context)
    {
    }

    public override async Task<MyShopUseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification>> Execute()
    {
        var result = await _driver.BrowseCouponsAsync();

        return new MyShopUseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification>(
            result,
            _context,
            (response, ctx) => new BrowseCouponsVerification(response, ctx));
    }
}
